pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -1734,15 +1734,36 @@ func buildTracingConfig(config *meshconfig.ProxyConfig) *hcm.HttpConnectionManag
 					Value: config.Tracing.MaxPathTagLength,
 				}
 		}
+		tracingCfg.CustomTags = buildCustomTags(config.Tracing.CustomTags)
 
-		if len(config.Tracing.CustomTags) != 0 {
-			tracingCfg.CustomTags = buildCustomTags(config.Tracing.CustomTags)
-		}
 	}
 
 	return tracingCfg
 }
 
+func buildCanonicalServiceTags() []*tracing.CustomTag {
+	tags := make([]*tracing.CustomTag, 0, 2)
+	tags = append(tags, &tracing.CustomTag{
+		Tag: "canonical_revision",
+		Type: &tracing.CustomTag_Environment_{
+			Environment: &tracing.CustomTag_Environment{
+				Name:         "CANONICAL_REVISION",
+				DefaultValue: "latest",
+			},
+		},
+	})
+	tags = append(tags, &tracing.CustomTag{
+		Tag: "canonical_service",
+		Type: &tracing.CustomTag_Environment_{
+			Environment: &tracing.CustomTag_Environment{
+				Name:         "CANONICAL_SERVICE",
```

Comment:
Where / who substitutes the real value?

Reply:
What do you mean by real value?  This tells Envoy to look up `CANONICAL_SERVICE` env and use it. The injector makes sure the proper value is set.